compiler/noirc_evaluator/src/ssa/opt/constant_evaluation.rs

```diff
@@ -0,0 +1,474 @@
+//! Evaluates calls to Brillig functions at compile time when all arguments are constant.
+//!
+//! This acts as a partial evaluator for Brillig entry points called from ACIR functions.
+//! For any Brillig function call where all arguments are constant, we interpret the
+//! function using the [SSA interpreter][crate::ssa::interpreter] and replace the call with the resulting constants.
+//!
+//! This improves constant folding across function boundaries, particularly for small utility
+//! Brillig functions that can be inlined
+//!
+//! Note: Only calls to Brillig functions from ACIR functions (Brillig entry points) are evaluated here.
+//! Most other constant evaluation occurs during [instruction simplification][crate::ssa::ir::dfg::simplify]
+//! and [constant folding][crate::ssa::opt::constant_folding].
+//! However, this constant evaluation is unable to remove calls to Brillig entry points with constant arguments
+//! as entry points are explicitly prevented from being inlined.  
+use fxhash::FxHashMap as HashMap;
+
+use crate::ssa::{
+    interpreter::{InterpreterOptions, value::Value},
+    ir::{
+        basic_block::BasicBlockId,
+        dfg::DataFlowGraph,
+        function::Function,
+        instruction::{Instruction, InstructionId},
+        types::Type,
+        value::{Value as IrValue, ValueId},
+    },
+    ssa_gen::Ssa,
+};
+
+impl Ssa {
+    /// See [`fold_constant_brillig_calls`][self] module for more information.
+    pub(crate) fn fold_constant_brillig_calls(mut self) -> Ssa {
+        // Gather constant evaluation results per function
+        let mut constant_evaluations = self
+            .functions
+            .iter()
+            .map(|(id, function)| (*id, function.constant_evaluation(&self)))
+            .collect::<HashMap<_, _>>();
+
+        // Apply constant folding for each function based on the interpreter results
+        for function in self.functions.values_mut() {
+            let Some(constant_evaluations) = constant_evaluations.remove(&function.id()) else {
+                continue;
+            };
```

Comment:
Is there a case where this returns `None`? Based on the above loop it seems every function ID will get a HashMap, or at least an empty one.
(maybe `constant_evaluation` could return `Option<HashMap<..>>`?)